### PR TITLE
Addition of new spec components to TKG Service vSphere cluster resources

### LINKED ISSUE #24: Ability to specify storage classes and default storage class for TKGS clusters
### Is your feature request related to a problem? Please describe.

Not related to a problem.  Not sure why this field is required.

### Describe the solution you'd like

The TMC UI allows me to specify a list of storage classes that I want deployed to the TKGS cluster and optionally one that I would like to set as the default storage class.  This is not available with the current tkg_service_vsphere model.  I would expect this in the settings block.

Here is a snippet of TMC yaml that includes this data.
```
spec:
  clusterGroupName: se-dpfeffer-dev
  tkgServiceVsphere:
    distribution:
      version: v1.21.2+vmware.1-tkg.1.ee25d55
    settings:
      network:
        pods:
          cidrBlocks:
          - 172.20.0.0/16
        services:
          cidrBlocks:
          - 10.96.0.0/16
      storage:
        classes:
        - tanzu
        defaultClass: tanzu
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Vasundharashukla**
1. **What this PR does / why we need it**: Added support for the optional parameters to TKGs and TKGm workload clusters

2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #24

3. **Additional information** 

Added `Volumes` and `Storage` to TKGs workload cluster spec. 
Added `APIServerPort` to TKGm vSphere workload cluster spec. 

4. **Special notes for your reviewer**

### CODE REVIEW

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go [hunk 1]
```diff
@@ -370,13 +442,92 @@ func flattenTKGSTopologyControlPlane(controlPlane *tkgservicevspheremodel.Vmware
...
+				Type:        schema.TypeString,
+				Description: "It is the directory where the volume device is to be mounted",
+				Optional:    true,
+			},
+			volumeNameKey: {
+				Type:        schema.TypeString,
+				Description: "It is the volume name",
+				Optional:    true,
+			},
+			pvcStorageClassKey: {
```
Comment (shobha2626) [inline internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:480]: How is this values populated? I don't see this in the UI. is it a computed value?

File: internal/resources/cluster/tkgvsphere/resource_tkgvsphere.go [hunk 2]
```diff
@@ -154,6 +154,12 @@ var tkgVsphereNetwork = &schema.Schema{
 	MaxItems:    1,
 	Elem: &schema.Resource{
 		Schema: map[string]*schema.Schema{
+			apiServerPortKey: {
+				Type:        schema.TypeInt,
+				Default:     apiServerPortDefaultValue,
```
Comment (shobha2626) [inline internal/resources/cluster/tkgvsphere/resource_tkgvsphere.go:159]: We should mention about the default value given in the doc also

File: internal/models/cluster/nodepool/nodepool_tkgs_volume.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import "github.com/go-openapi/swag"
+
+// VmwareTanzuManageV1alpha1CommonClusterTKGServiceVsphereVolume TKGServiceVsphereVolume defines a Persistent Volume Claim attached for the workload cluster.
+//
+// swagger:model vmware.tanzu.manage.v1alpha1.common.cluster.TKGServiceVsphereVolume
+type VmwareTanzuManageV1alpha1CommonClusterTKGServiceVsphereVolume struct {
+
+	// Volume capacity is in gib.
+	Capacity float32 `json:"capacity,omitempty"`
```
Comment (shobha2626) [inline internal/models/cluster/nodepool/nodepool_tkgs_volume.go:16]: I don't see the schema for these optional entries we can enter for the Nodepools, please add the same. We should provide this for both the TKGs and TKGm.

Note: Adding node pool is supported for TKGm workload clusters with version greater then 1.4.1

**Review by shobha2626 — CHANGES_REQUESTED** [review]

**Comment by shobha2626** [comment]
Please fill the description of the PR

File: internal/models/cluster/nodepool/nodepool_tkgs_volume.go (hunk 3 above)
Comment (shobha2626) [reply internal/models/cluster/nodepool/nodepool_tkgs_volume.go:16]: Please check the same for AWS workload cluster

**Comment by shobha2626** [comment]
We should regenerate the document, so that the new fields are also captured

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go [hunk 4]
```diff
@@ -110,12 +111,38 @@ var network = &schema.Schema{
 	},
 }
 
+var storage = &schema.Schema{
+	Type:        schema.TypeList,
+	Description: "StorageSettings specifies storage-related settings for the cluster",
+	Optional:    true,
+	Elem: &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			classesKey: {
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:120]: Changed to the `defaultStorageClassKey`

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go (hunk 1 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:480]: It is the Storage Class value under Volumes in the UI

File: internal/resources/cluster/tkgvsphere/resource_tkgvsphere.go (hunk 2 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/tkgvsphere/resource_tkgvsphere.go:159]: The generated docs contain the default value.

File: internal/models/cluster/nodepool/nodepool_tkgs_volume.go (hunk 3 above)
Comment (Vasundharashukla) [reply internal/models/cluster/nodepool/nodepool_tkgs_volume.go:16]: As discussed in our call, the `volumes` spec has been added to the node pools. Also, TKGm vSphere and AWS doesn't have these fields in the spec.

File: docs/guides/tanzu-mission-control_cluster.md [hunk 5]
```diff
@@ -78,8 +78,14 @@ resource "tanzu-mission-control_cluster" "create_tkgs_workload" {
       topology {
         control_plane {
           class             = "<class>"        // Required
-          storage_class     = "<storage-class" // Required
+          storage_class     = "<storage-class>" // Required
           high_availability = false            // Default: false
+          volumes {
+            capacity          = volume-capacity
```
Comment (gsadhani) [inline docs/guides/tanzu-mission-control_cluster.md:84]: `<volume-capacity>`

File: docs/guides/tanzu-mission-control_cluster.md [hunk 6]
```diff
@@ -93,6 +99,12 @@ resource "tanzu-mission-control_cluster" "create_tkgs_workload" {
             tkg_service_vsphere {
               class         = "<class>"         // Required
               storage_class = "<storage-class>" // Required
+              volumes {
+                capacity          = volume-capacity
```
Comment (gsadhani) [inline docs/guides/tanzu-mission-control_cluster.md:103]: `<volume-capacity>`

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gsadhani) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:120]: classesKey => storageClassKey is still pending I guess.

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go [hunk 7]
```diff
@@ -370,13 +442,92 @@ func flattenTKGSTopologyControlPlane(controlPlane *tkgservicevspheremodel.Vmware
...
+			pvcStorageClassKey: {
+				Type:        schema.TypeString,
+				Description: "This is the storage class for PVC which in case omitted, default storage class will be used for the disks",
+				Optional:    true,
+			},
+		},
+	},
+}
+
+func expandTKGSVolumes(data interface{}) (volumes *nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGServiceVsphereVolume) {
```
Comment (gsadhani) [inline internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:489]: nit: `volume`

File: internal/models/cluster/nodepool/nodepool_tkgs_volume.go (hunk 3 above)
Comment (shobha2626) [reply internal/models/cluster/nodepool/nodepool_tkgs_volume.go:16]: TKGm vSphere above 1.4.1 doesn't have these fields?

**Review by gsadhani — COMMENTED** [review]
LGTM overall.

File: docs/index.md [hunk 8]
```diff
@@ -58,7 +58,4 @@ provider "tanzu-mission-control" {
 
 - **endpoint** (String)
 - **vmw_cloud_api_token** (String, Sensitive)
-
-### Optional
```
Comment (shobha2626) [inline docs/index.md:62]: Retain this

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go [hunk 9]
```diff
@@ -75,6 +75,7 @@ var settings = &schema.Schema{
 	Elem: &schema.Resource{
 		Schema: map[string]*schema.Schema{
 			networkKey: network,
+			storageKey: storage,
```
Comment (shobha2626) [inline internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:78]: This field is empty right? We concluded to let it be in the model but not spec

File: docs/data-sources/cluster.md [hunk 10]
```diff
@@ -117,7 +117,7 @@ Optional:
 - **cluster_group** (String) Name of the cluster group to which this cluster belongs
 - **proxy** (String) Optional proxy name is the name of the Proxy Config to be used for the cluster
 - **tkg_service_vsphere** (Block List, Max: 1) The Tanzu Kubernetes Grid Service (TKGs) cluster spec (see [below for nested schema](#nestedblock--spec--tkg_service_vsphere))
-- **tkg_vsphere** (Block List, Max: 1) The Tanzu Kubernetes Grid (TKGm) cluster spec (see [below for nested schema](#nestedblock--spec--tkg_vsphere))
+- **tkg_vsphere** (Block List, Max: 1) The Tanzu Kubernetes Grid (TKGm) VSphere cluster spec (see [below for nested schema](#nestedblock--spec--tkg_vsphere))
```
[the remark this answers is not in the document]
Comment (shobha2626) [reply docs/data-sources/cluster.md:120]: I see it is captured in the PR https://github.com/vmware/terraform-provider-tanzu-mission-control/pull/34

File: internal/models/cluster/nodepool/nodepool_tkgs_volume.go (hunk 3 above)
Comment (shobha2626) [reply internal/models/cluster/nodepool/nodepool_tkgs_volume.go:16]: Okay, we don not have such fields in the model itself 👍

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go (hunk 9 above)
Comment (shobha2626) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go]: The renamed files should also be reverted after removing this

File: docs/data-sources/cluster.md [hunk 11]
```diff
@@ -143,31 +143,41 @@ Required:
...
-- **services** (Block List, Min: 1) Service CIDR for kubernetes services defaults to 10.96.0.0/12 (see [below for nested schema](#nestedblock--spec--tkg_service_vsphere--settings--network--services))
+- **pods** (Block List, Min: 1) Pod CIDR for Kubernetes pods defaults to 192.168.0.0/16 (see [below for nested schema](#nestedblock--spec--tkg_service_vsphere--settings--storage--pods))
+- **services** (Block List, Min: 1) Service CIDR for kubernetes services defaults to 10.96.0.0/12 (see [below for nested schema](#nestedblock--spec--tkg_service_vsphere--settings--storage--services))
 
-<a id="nestedblock--spec--tkg_service_vsphere--settings--network--pods"></a>
-### Nested Schema for `spec.tkg_service_vsphere.settings.network.pods`
+<a id="nestedblock--spec--tkg_service_vsphere--settings--storage--pods"></a>
+### Nested Schema for `spec.tkg_service_vsphere.settings.storage.pods`
 
 Required:
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md]: @gsadhani It is a required field as the creation is not possible without entering this value.

File: docs/data-sources/cluster.md [hunk 12]
```diff
@@ -143,31 +143,41 @@ Required:
...
 
 - **cidr_blocks** (List of String) CIDRBlocks specifies one or more ranges of IP addresses
 
 
-<a id="nestedblock--spec--tkg_service_vsphere--settings--network--services"></a>
-### Nested Schema for `spec.tkg_service_vsphere.settings.network.services`
+<a id="nestedblock--spec--tkg_service_vsphere--settings--storage--services"></a>
+### Nested Schema for `spec.tkg_service_vsphere.settings.storage.services`
 
 Required:
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md]: @gsadhani It is a required field as the creation is not possible without entering this value.

File: docs/guides/tanzu-mission-control_cluster.md (hunk 5 above)
Comment (Vasundharashukla) [reply docs/guides/tanzu-mission-control_cluster.md:84]: Capacity is a float type variable, we cannot have it in the format `<volume-capacity>` or `"<volume-capacity>"`

File: examples/usecases/usecase.tf [hunk 13]
```diff
@@ -199,7 +217,7 @@ resource "tanzu-mission-control_cluster" "create_tkg_vsphere_cluster" {
           }
 
           info {
-            name        = "default-nodepool"
+            name        = "md-0"
```
Comment (shobha2626) [inline examples/usecases/usecase.tf:220]: The creation fails with `default-node pool` name is it? It seems to be computed for TKGm vSphere and AWS (default node pool creation)

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go (hunk 7 above)
Comment (Vasundharashukla) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go]: Done

File: docs/data-sources/cluster.md [hunk 14]
```diff
@@ -143,31 +143,41 @@ Required:
 
 - **network** (Block List, Min: 1, Max: 1) Network Settings specifies network-related settings for the cluster (see [below for nested schema](#nestedblock--spec--tkg_service_vsphere--settings--network))
 
+Optional:
+
+- **storage** (Block List) StorageSettings specifies storage-related settings for the cluster (see [below for nested schema](#nestedblock--spec--tkg_service_vsphere--settings--storage))
+
 <a id="nestedblock--spec--tkg_service_vsphere--settings--network"></a>
-### Nested Schema for `spec.tkg_service_vsphere.settings.network`
```
[the remark this answers is not in the document]
Comment (Vasundharashukla) [reply docs/data-sources/cluster.md]: Addressed

File: examples/usecases/usecase.tf (hunk 13 above)
Comment (shobha2626) [reply examples/usecases/usecase.tf:220]: Node Pool Name
1. TKGm vSphere and AWS - computed
2. TKGs - required

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go [hunk 15]
```diff
@@ -424,11 +516,11 @@ var info = &schema.Schema{
 	MaxItems:    1,
 	Elem: &schema.Resource{
 		Schema: map[string]*schema.Schema{
-			clusterNameKey: {
+			nodepoolNameKey: {
```
Comment (shobha2626) [inline internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:519]: Can we do similar change for TKGs as well?

Comment (shobha2626) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:519]: Its super nit comment, these constants are pointing to `name` anyways

**Review by shobha2626 — DISMISSED** [review]
LGTM

**Comment by shobha2626** [comment]
Squash the commits into one

**Comment by gsadhani** [comment]
LGTM

**Review by gsadhani — DISMISSED** [review]

**Review by shobha2626 — DISMISSED** [review]

File: internal/resources/cluster/resource_cluster.go [hunk 16]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright © 2021 VMware, Inc. All Rights Reserved.
+Copyright © 2022 VMware, Inc. All Rights Reserved.
```
Comment (shreyassreenivas) [inline internal/resources/cluster/resource_cluster.go:2]: this should not be changed

File: internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go [hunk 17]
```diff
@@ -424,11 +516,11 @@ var info = &schema.Schema{
...
 				Type:        schema.TypeString,
 				Description: "Name of the nodepool",
 				Default:     defaultNodePoolName,
-				Optional:    true,
+				Required:    true,
```
Comment (shreyassreenivas) [inline internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:523]: Default and Required cannot be set . 

`// If Required is true above, then Default cannot be set. DefaultFunc
	// can be set with Required. If the DefaultFunc returns nil, then there
	// will be no default and the user will be asked to fill it in.`

Comment (shobha2626) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go:523]: Please revert this change and regenerate the doc

Comment (Vasundharashukla) [reply internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go]: Thanks for pointing this out @shreyassreenivas, we were missing on this as we were running the acceptance test for AWS resource and checking only the spec of that resource.

**Review by gsadhani — APPROVED** [review]

**Review by shobha2626 — APPROVED** [review]

File: internal/resources/cluster/resource_cluster.go (hunk 16 above)
Comment (shreyassreenivas) [reply internal/resources/cluster/resource_cluster.go]: This was not addressed and you marked it resolved ?

cc: @shobha2626

Comment (shobha2626) [reply internal/resources/cluster/resource_cluster.go]: @Vasundharashukla please take care in the PR - https://github.com/vmware/terraform-provider-tanzu-mission-control/pull/34

**Pull request merged**